fl4health/clients/moon_client.py

```diff
@@ -25,33 +22,58 @@ def __init__(
+            device (torch.device): Device to be used for training.
+            loss_meter_type (LossMeterType, optional): Type of loss meter to be used. Defaults to
+                LossMeterType.AVERAGE.
+            checkpointer (Optional[ClientCheckpointModule], optional): Checkpointer module defining when and how to
+                do checkpointing during client-side training. No checkpointing is done if not provided. Defaults to
+                None.
+            temperature (float, optional): Temperature used in the calculation of the contrastive loss.
+                Defaults to 0.5.
+            contrastive_weight (float, optional): Weight placed on the contrastive loss function. Referred to as mu
+                in the original paper. Defaults to 1.0.
+            len_old_models_buffer (int, optional): Number of old models to be stored for computation in the
+                contrastive learning loss function. Defaults to 1.
+        """
         super().__init__(
             data_path=data_path,
             metrics=metrics,
             device=device,
             loss_meter_type=loss_meter_type,
             checkpointer=checkpointer,
         )
-        self.cos_sim = torch.nn.CosineSimilarity(dim=-1).to(self.device)
-        self.ce_criterion = torch.nn.CrossEntropyLoss().to(self.device)
-        self.contrastive_weight = contrastive_weight
         self.temperature = temperature
+        self.contrastive_weight = contrastive_weight
+        if self.contrastive_weight == 0:
+            log(WARNING, "Contrastive loss weight is set to 0, thus Contrastive loss will not be computed.")
+        self.contrastive_loss_function = ContrastiveLoss(self.device, temperature=temperature)
 
-        # Saving previous local models and global model at each communication round to compute contrastive loss
+        # Saving previous local models and a global model at each communication round to compute contrastive loss
         self.len_old_models_buffer = len_old_models_buffer
         self.old_models_list: list[torch.nn.Module] = []
         self.global_model: Optional[torch.nn.Module] = None
 
     def predict(self, input: TorchInputType) -> Tuple[Dict[str, torch.Tensor], Dict[str, torch.Tensor]]:
         """
-        Computes the prediction(s) and features of the model(s) given the input.
+        Computes the prediction(s) and features of the model(s) given the input. This function also produces the
+        necessary features from the global_model (aggregated model from server) and old_models (previous client-side
+        optimized models) in order to be able to compute the appropriate contrastive loss.
 
         Args:
             input (TorchInputType): Inputs to be fed into the model. TorchInputType is simply an alias
-            for the union of torch.Tensor and Dict[str, torch.Tensor].
+            for the union of torch.Tensor and Dict[str, torch.Tensor]. Here, the Moon models require input to
```

Comment:
nit: In the documentation, we have been going with the MOON as opposed to Moon it seems like. Just wanted to flag that

Reply:
Yup, will make it consistent. I'll also do a lookup to see if we have that inconsistency elsewhere.